src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -461,96 +556,106 @@ public void OnResponseData(ReadOnlySpan<byte> buffer, bool endStream)
                 }
             }
 
-            public void OnAbort(Exception abortException, bool canRetry = false)
+            public void OnReset(Exception resetException, bool canRetry = false)
             {
-                bool signalWaiter;
+                if (NetEventSource.IsEnabled) Trace($"{nameof(resetException)}={resetException}");
+
                 lock (SyncObject)
                 {
-                    if (NetEventSource.IsEnabled) Trace($"{nameof(abortException)}={abortException}");
-
-                    if (_disposed || _state == StreamState.Aborted)
+                    // It's possible we could be called twice, e.g. we receive a RST_STREAM and then the whole connection dies
+                    // before we have a chance to process cancellation and tear everything down. Just ignore this.
+                    if (_resetException != null)
                     {
                         return;
                     }
 
-                    // We should not retry request which have started being processed since the behavior might be unpredictable
-                    // I.e. some action might have been taken based on the received data.
-                    // We will bubble the exception and let user decide what to do.
-                    bool isRetriable = _state == StreamState.ExpectingStatus || _state == StreamState.ExpectingHeaders;
-                    Interlocked.CompareExchange(ref _abortException, abortException, null);
-                    _state = StreamState.Aborted;
-                    _canRetry = canRetry && isRetriable;
+                    // If the server told us the request has not been processed (via Last-Stream-ID on GOAWAY),
+                    // but we've already received some response data from the server, then the server lied to us.
+                    // In this case, don't allow the request to be retried.
+                    if (canRetry && _responseProtocolState != ResponseProtocolState.ExpectingStatus)
+                    {
+                        canRetry = false;
+                    }
 
-                    signalWaiter = _hasWaiter;
-                    _hasWaiter = false;
+                    _resetException = resetException;
+                    _canRetry = canRetry;
                 }
 
-                if (signalWaiter)
-                {
-                    _waitSource.SetResult(true);
-                }
+                CancelRequestBody();
+                CancelResponseBody();
```

Comment:
CancelResponseBody calls CancelRequestBody.  Do we need both here?

Reply:
I've been thinking I may combine these into one routine anyway, just Cancel(). As you point out, CancelResponseBody calls CancelRequestBody anyway. And we only call CancelRequestBody without calling CancelResponseBody in cases where the response body is already completed (I think -- will verify), so in that case cancelling the response body is a no-op anyway. Will look at this further.